### ISSUE TITLE
Bad (not reverse of model introduction) JOIN order causes runtime error

### ISSUE THREAD

**Issue by bitemyapp** [bitemyapp/esqueleto#49]
This needs to be documented so that we understand precisely what causes it and what can be done to fix it.

**Comment by FintanH** [comment]
I found the author explaining why the reverse order is needed in the documentation in `Language.hs`.

> The order is reversed in order to improve composability. For example, consider query1 and query2 below:
> 
> ``` Haskell
> let query1 =
>       from $ \(foo `InnerJoin` bar) -> do
>         on (foo ^. FooId ==. bar ^. BarFooId)
>     query2 =
>       from $ \(mbaz `LeftOuterJoin` quux) -> do
>         return (mbaz ?. BazName, quux)
>     test1 =      (,) <$> query1 <*> query2
>     test2 = flip (,) <$> query2 <*> query1
> ```
> If the order was not reversed, then test2 would be broken: query1's on would refer to query2's LeftOuterJoin.

I'm trying to tease out why this breaks with his example, looking at the types, but maybe you can spot the reason quicker than I can.

**Comment by qrilka** [comment]
While porting significant amount of queries (BTW @bitemyapp you've seen many of them I think :) )  from `mysql-simple` to `persistent` + `esqeleto` this have bitten me a bit - it just gives SQL error, not even exception from the library.
@FintanH it breaks because joins and "on" conditions get mismatched, i.e. condition from `query1` gets used for `LeftOuterJoin` between `mbaz` and `quux` while it talks about `foo` and `bar`.

**Comment by MaxGabriel** [comment]
I know this isn't getting fixed in the current design, but has anyone come up with any workaround or way to programmatically check this by any chance? It's been the cause of two bugs for us

**Comment by bitemyapp** [comment]
@MaxGabriel The closest thing to a workaround would be [dovetail](https://github.com/frontrowed/dovetail) but I am pretty sure Front Row Ed no longer uses it.

Otherwise, I need time to work on [bluetick](https://github.com/bitemyapp/bluetick).

**Comment by parsonsmatt** [comment]
I'm exploring this to see if there's something that can be done to fix this up.

First, let's look at the offending function, `on`:

```haskell
on :: SqlExpr (Value Bool) -> SqlQuery ()
on expr = Q $ W.tell mempty { sdFromClause = [OnClause expr] }
```

The definition of `SqlQuery`:

```haskell
-- | SQL backend for @esqueleto@ using 'SqlPersistT'.
newtype SqlQuery a =
  Q { unQ :: W.WriterT SideData (S.State IdentState) a }
```

The `SideData` type is a collection of monoids that are `<>`ed together to construct the complete SQL query. One of those monoids is the `sdFromClause :: [FromClause]` type. The `on` function accepts a `Bool`-valued expression, which means we're going to need to introspect on the `SqlExpr` bit of it - that's going to be... annoying. Can we work around it?

First, let's see what a `FromClause` is all about.

```haskell
-- | A part of a @FROM@ clause.
data FromClause =
    FromStart Ident EntityDef
  | FromJoin FromClause JoinKind FromClause (Maybe (SqlExpr (Value Bool)))
  | OnClause (SqlExpr (Value Bool))
```

We have the beginning, then a `Join` (a binary tree of `FromClause` annotated with a `JoinKind` and a `Maybe (SqlExpr (Vale Bool))`, and then an `OnClause (SqlExpr (Value Bool))`. An ordinary SQL query will have a single `FromClause` that is `FromStart` with the identifier for the entity and the entity definition.

The `OnClause` is meant to be eliminated and absorbed into the `FromJoin` tree. All of the `FromJoin`s will start with a `Nothing` value in the `Maybe (SqlExpr (Value Bool))` field, and will later have those expressions filled in.

The function that builds up the SQL text from the `[FromClause]` is `makeFrom`.

```haskell
makeFrom :: IdentInfo -> Mode -> [FromClause] -> (TLB.Builder, [PersistValue])
makeFrom _    _    [] = mempty
makeFrom info mode fs = ret
  where
    ret = case collectOnClauses fs of
            Left expr -> throw $ mkExc expr
            Right fs' -> keyword $ uncommas' (map (mk Never) fs')
```

We get a SQL error, and not an `esqueleto` error, which means that `collectOnClauses` is returning a `Right`. That function basically just checks to ensure that we have an equal amount of `Join`s and `on` calls.

Let's look at `collectOnClauses` - this one is succeeding in this case.

```haskell
collectOnClauses :: [FromClause] -> Either (SqlExpr (Value Bool)) [FromClause]
collectOnClauses = go []
  where
    go []  (f@(FromStart _ _):fs) = fmap (f:) (go [] fs) -- fast path
    go acc (OnClause expr    :fs) = do
      a <- findMatching acc expr
      go a fs
    go acc (f:fs)                 = go (f:acc) fs
    go acc []                     = return $ reverse acc

   -- snip
```

This is written in a manually recursive style, so it's a little hard to follow IMO. The only interesting case is `OnClause expr` - everything else is a mostly straightforward iteration through the list. Once we're done, we reverse the `acc`. So the first case is a way of saying "Put the `FromStart` at the beginning of the result list," but the *rest* of the list will be reversed.

Let's look at `findMatching`, which does the interesting part in this:

```haskell
    findMatching
      :: [FromClause]
      -> SqlExpr (Value Bool)
      -> Either (SqlExpr (Value Bool)) [FromClause]
    findMatching fromClauses expr =
      case fromClauses of
        f : acc ->
          case tryMatch expr f of
            Just f' -> return (f' : acc)
            Nothing -> (f:) <$> findMatching acc expr
        [] ->
          Left expr
```

More manual recursion. We're iterating down the `fromClauses` list - if it's empty, then we fail out with the `Left` expression. If it's not empty, then we `tryMatch` on the list. If the match succeeds with `Just f'`, then we're early-terminating and returning `f'` in place of `acc`. If `tryMatch` fails, then we put `f` back in place in the result list and recurse.

The logic here is branching on `tryMatch`,so let's look at that:

```haskell
    tryMatch :: SqlExpr (Value Bool) -> FromClause -> Maybe FromClause
    tryMatch expr fromClause =
      case fromClause of
        FromJoin l k r onClause ->
          matchR `mplus` matchC `mplus` matchL -- right to left
            where
              matchR = (\r' -> FromJoin l k r' onClause) <$> tryMatch expr r
              matchL = (\l' -> FromJoin l' k r onClause) <$> tryMatch expr l
              matchC =
                case onClause of
                  Nothing
                    | k /= CrossJoinKind -> return (FromJoin l k r (Just expr))
                    | otherwise -> Nothing
                  Just _ ->
                    Nothing
        _ ->
          Nothing
```

Hooray for manual recursion! I guess I signed up for this when I agreed to write Haskell for a living. We deconstruct the `fromClause` and we only care about `FromJoin`s here - anything else causes a `Nothing` result (which means we'd try the next one in `findMatching`).

`matchR` and `matchL` are relatively simple recursions, so we can pretty much focus our attention on `matchC`. The *only* time it returns `Just` is if `onClause` is `Nothing` *and* if `k` is not `CrossJoin`. A `Just` value here replaces the `onClause` with `Just expr`. So this is *kinda* like a way of saying: "Give me a `FromClause` and an `expr :: SqlExpr (Value Bool)`, and I'll figure out a place to put the `OnClause`."

It tries the rightmost bit first, and if that doesn't work, it tries the current side, and if that doesn't work, it tries the leftmost bit. So it really does just look at the rightmost part of the tree that isn't currently filled in and place it there.

Once we've successfully taken the `OnClause`s and shoved them into the "corresponding" `FromClause`s, we get back to `makeFrom`. Now we do `map (mk Never fs')`. So let's look at that.

```haskell
    mk _     (FromStart i def) = base i def
    mk paren (FromJoin lhs kind rhs monClause) =
      first (parensM paren) $
      mconcat [ mk Never lhs
              , (fromKind kind, mempty)
              , mk Parens rhs
              , maybe mempty makeOnClause monClause
              ]
    mk _ (OnClause _) = throw (UnexpectedCaseErr MakeFromError)
```

`mk` is relatively simple. We assume that we have no further `ON` clauses in the data structure, because they've been absorbed into a `FromClause`s. Again it recurses through the binary tree, folding it into the `(TLB.Builder, [PersistVal])` result.

Aside from some formatting stuff, that's totally it.

So, if we're going to make this smarter, in a backwards-compatible way, I think we're going to need to be smarter about `tryMatch`. Currently, it just puts the `OnClause expr` in the first available slot. Instead, I think we're going to need to have it search somehow. It's going to be tricky, because the `expr` has erased all type information: `SqlExpr (Value Bool)`.

A ground-up redesign would have something like:

```haskell
on 
  :: (PersistEntity left, PersistEntity right, PersistField joinColumn)
  => (SqlExpr (Entity left) -> SqlExpr (Value joinColumn))
  -> (SqlExpr (Entity right) -> SqlExpr (Value joinColumn))
  -> SqlQuery ()
```

and would be used like:

```haskell
select $
from $ \(user `InnerJoin` blog) -> do
on (^. UserId) (^. BlogAuthor)
pure (user, blog)
```

Then, `on` could take the `left` and `right` type variables and use them to properly associate the joins.

But this isn't a perfect world. So we need to dig into the `SqlExpr (Value Bool)` and see what we can get.

Virtually ever use of `on` in the work codebase is the form:

```haskell
  on (x ^. EntityId ==. y ^. OtherEntityId)
```

The outermost expression is `==.`, so what's that look like? 

```haskell
(==.) :: PersistField typ => SqlExpr (Value typ) -> SqlExpr (Value typ) -> SqlExpr (Value Bool)
(==.) = unsafeSqlBinOpComposite " = " " AND "
```

Dig:

```haskell
unsafeSqlBinOpComposite :: TLB.Builder -> TLB.Builder -> SqlExpr (Value a) -> SqlExpr (Value b) -> SqlExpr (Value c)
unsafeSqlBinOpComposite op _ a@(ERaw _ _) b@(ERaw _ _) = unsafeSqlBinOp op a b
unsafeSqlBinOpComposite op sep a b = ERaw Parens $ compose (listify a) (listify b)
  -- a bunch of stuff to deal with composite keys
``

Dig:

```haskell
unsafeSqlBinOp :: TLB.Builder -> SqlExpr (Value a) -> SqlExpr (Value b) -> SqlExpr (Value c)
unsafeSqlBinOp op (ERaw p1 f1) (ERaw p2 f2) = ERaw Parens f
  where
    f info = let (b1, vals1) = f1 info
                 (b2, vals2) = f2 info
             in ( parensM p1 b1 <> op <> parensM p2 b2
                , vals1 <> vals2 )
unsafeSqlBinOp op a b = unsafeSqlBinOp op (construct a) (construct b)
    where construct :: SqlExpr (Value a) -> SqlExpr (Value a)
          construct (ERaw p f)        = ERaw Parens $ \info ->
            let (b1, vals) = f info
                build ("?", [PersistList vals']) =
                  (uncommas $ replicate (length vals') "?", vals')
                build expr = expr
             in  build (parensM p b1, vals)
          construct (ECompositeKey f) =
            ERaw Parens $ \info -> (uncommas $ f info, mempty)
```

So it takes the `ECompositeKey`, converts it to `ERaw` if necessary, and then processes it. We're probably dealing with `ERaw` for the basic case, so let's just look at that. We construct those `ERaw` values using the `^.` operator, so let's look at that definition:

```haskell
(^.)
  :: forall typ val. (PersistEntity val, PersistField typ)
  => SqlExpr (Entity val)
  -> EntityField val typ
  -> SqlExpr (Value typ)
EEntity ident ^. field
  | isComposite = ECompositeKey $ \info ->  dot info <$> compositeFields pdef
  | otherwise   = ERaw Never    $ \info -> (dot info  $  persistFieldDef field, [])
  where
    isComposite = isIdField field && hasCompositeKey ed
    dot info x  = useIdent info ident <> "." <> fromDBName info (fieldDB x)
    ed          = entityDef $ getEntityVal (Proxy :: Proxy (SqlExpr (Entity val)))
    Just pdef   = entityPrimary ed
```

Okay, so we don't have any `PersistValue`s to worry about, just the `TLB.Builder`. And that gets constructed by doing `dot info (persistFieldDef field)`, which decomposes to `useIdent info ident <> "." <> fromDBName info (fieldDB field)`. It may be possible to render the text of the `SqlExpr (Value Bool)` and dig out the table names, and it *may* be possible to go from there to joining the right things. Welp.

`useIdent` and `fromDBName` require the `info`, so let's look at those:

```haskell
-- | Information needed to escape and use identifiers.
type IdentInfo = (SqlBackend, IdentState)

useIdent :: IdentInfo -> Ident -> TLB.Builder
useIdent info (I ident) = fromDBName info $ DBName ident

fromDBName :: IdentInfo -> DBName -> TLB.Builder
fromDBName (conn, _) = TLB.fromText . connEscapeName conn
```

Nice, so we don't actually need an `IdentState` at all to accomplish this, just the `SqlBackend`.
Since the `SqlBackend` should be mostly stable throughout an application lifetime, this should work out.

Let's write a function to render a `SqlExpr (Value x)`.

```haskell
renderExpr :: Monad m => SqlExpr (Value a) -> SqlPersistT m T.Text
renderExpr e = case e of
  ERaw parens mkBuilderValues -> do
    sqlBackend <- R.ask
    let (builder, _) = mkBuilderValues (sqlBackend, initialIdentState)
    pure (builderToText builder)
  ECompositeKey _ ->
    error "don't care"
```

I put this in `SqlPersistT` because that'll make it easier to use in blocks we're doing.
Let's test it:

```haskell
  describe "renderExpr" $ do
    it "renders a value" $ do
      run $ do
        expr <- EI.renderExpr $
          EI.EEntity (EI.I "user") ^. PersonId
          ==. EI.EEntity (EI.I "blog_post") ^. BlogPostAuthorId
        liftIO $ expr `shouldBe` "\"user\".\"id\" == \"blog_post\".\"author_id\""

```

This passes.
So can we parse this out from this decomposed string into the types we need to make it work out?

**Comment by parsonsmatt** [comment]
# Part Two

Let's look at how the `sdFromClause :: [FromClause]` get built. The pattern is to write:

```haskell
addFromClause :: FromClause -> SqlQuery ()
addFromClause fc = Q $ W.tell mempty { sdFromClause = [fc] }
```

This is called with `on` and `fromFinish`.
`fromFinish` is called a bunch in composition with `fromPreprocess`, in instance slike:

```haskell
instance ( FromPreprocess (SqlExpr (Entity val))
         ) => From (SqlExpr (Entity val)) where
  from_ = fromPreprocess >>= fromFinish
```

`fromPreprocess` is a class method:

```haskell
-- | (Internal) Class that implements the @JOIN@ 'from' magic
-- (see 'fromStart').
class FromPreprocess a where
  fromPreprocess :: SqlQuery (SqlExpr (PreprocessedFrom a))
```

Well, we should probably look at `fromStart` at some point. But for right now, I'm going to focus on the instances here:

```haskell
instance ( PersistEntity val
         , BackendCompatible SqlBackend (PersistEntityBackend val)
         ) => FromPreprocess (SqlExpr (Entity val)) where
  fromPreprocess = fromStart

instance ( PersistEntity val
         , BackendCompatible SqlBackend (PersistEntityBackend val)
         ) => FromPreprocess (SqlExpr (Maybe (Entity val))) where
  fromPreprocess = fromStartMaybe

instance ( FromPreprocess a
         , FromPreprocess b
         , IsJoinKind join
         ) => FromPreprocess (join a b) where
  fromPreprocess = do
    a <- fromPreprocess
    b <- fromPreprocess
    fromJoin a b
```

So we've got `fromStart`, `fromStartMaybe`, and `fromJoin` to look at.

```haskell
-- | (Internal) Start a 'from' query with an entity. 'from'
-- does two kinds of magic using 'fromStart', 'fromJoin' and
-- 'fromFinish':
--
--   1.  The simple but tedious magic of allowing tuples to be
--   used.
--
--   2.  The more advanced magic of creating @JOIN@s.  The
--   @JOIN@ is processed from right to left.  The rightmost
--   entity of the @JOIN@ is created with 'fromStart'.  Each
--   @JOIN@ step is then translated into a call to 'fromJoin'.
--   In the end, 'fromFinish' is called to materialize the
--   @JOIN@.
fromStart
  :: ( PersistEntity a
     , BackendCompatible SqlBackend (PersistEntityBackend a) )
  => SqlQuery (SqlExpr (PreprocessedFrom (SqlExpr (Entity a))))
fromStart = x
  where
    x = do
      let ed = entityDef (getVal x)
      ident <- newIdentFor (entityDB ed)
      let ret   = EEntity ident
          from_ = FromStart ident ed
      return (EPreprocessedFrom ret from_)
    getVal :: SqlQuery (SqlExpr (PreprocessedFrom (SqlExpr (Entity a)))) -> Proxy a
    getVal = const Proxy
```

The comments say we construct the values from right to left, so we get `FromStart` with the rightmost entity. The `FromClause` gets the `ident`, which is a fresh ident for the entity. This presents our first problem: if the identifier is not fresh, then we won't be able to easily parse it back out. However, we can find a *partial* match in the text and then try whatever mutations `newIdentFor` performs.

`fromStartMaybe` is kind of boring, just massages the type around, so let's get into `fromJoin`.

```haskell
-- | (Internal) Do a @JOIN@.
fromJoin
  :: IsJoinKind join
  => SqlExpr (PreprocessedFrom a)
  -> SqlExpr (PreprocessedFrom b)
  -> SqlQuery (SqlExpr (PreprocessedFrom (join a b)))
fromJoin (EPreprocessedFrom lhsRet lhsFrom)
         (EPreprocessedFrom rhsRet rhsFrom) = Q $ do
  let ret   = smartJoin lhsRet rhsRet
      from_ = FromJoin lhsFrom             -- LHS
                       (reifyJoinKind ret) -- JOIN
                       rhsFrom             -- RHS
                       Nothing             -- ON
  return (EPreprocessedFrom ret from_)
```

A decent bit of the polymorphic magic is to allow this to work with `InnerJoin`, `LeftOuterJoin`, etc. `fromJoin` is called with the result of calling `fromPreprocess` on the left and right hand sides and combining them, and this should bottom out with `fromStart` eventually.

Let's do some equational reasoning and figure out what *actually* happens when we have something like:

```haskell
from $ \(a `InnerJoin` b `InnerJoin` c) -> ...
```

`from :: (From a) => (a -> SqlQuery b) -> SqlQuery b`. So we need a `From a` instance. But wait, what's the fixity on those constructors?

```haskell
infixl 2 `InnerJoin`, `CrossJoin`, `LeftOuterJoin`, `RightOuterJoin`, `FullOuterJoin`
```

Okay, so they associate to the left, so ``a `InnerJoin` b `InnerJoin` c`` evaluates to `InnerJoin (InnerJoin a b) c`. So we're looking for an instance that'll give us:

```haskell
From (InnerJoin (InnerJoin a b) c)
```

We have one:

```haskell
instance ( FromPreprocess (InnerJoin a b)
         ) => From (InnerJoin a b) where
  from_ = fromPreprocess >>= fromFinish
```

Now we need to look at the instance for `FormPreprocess`:

```haskell
instance ( FromPreprocess a
         , FromPreprocess b
         , IsJoinKind join
         ) => FromPreprocess (join a b) where
  fromPreprocess = do
    a <- fromPreprocess
    b <- fromPreprocess
    fromJoin a b
```

We do have an instance for `IsJoinKind InnerJoin`, and it's irrelevant, so I won't look at it here. `fromPreprocess` is used polymorphically here, so we really have the annotated version:

```haskell
instance ( FromPreprocess a
         , FromPreprocess b
         , FromPreprocess c
         ) => FromPreprocess (InnerJoin (InnerJoin a b) c) where
  fromPreprocess = do
    a <- fromPreprocess :: _ (InnerJoin a b)
    b <- fromPreprocess :: _ c
    fromJoin a b
```

`fromStart` is the implementation that gets picked for `fromPreprocess :: _ c`, and for `fromPreprocess :: _ (InnerJoin a b)`, it'll pretty much recurse. So the final "filled in" implementation is:

```haskell
fromPreprocess = do
  ab <- do
    a <- fromStart
    b <- fromStart
    fromJoin a b
  c <- fromStart
  fromJoin ab c
-- , or, equivalently,
fromPreprocess = do
  a <- fromStart
  b <- fromStart
  ab <- fromJoin a b
  c <- fromStart
  fromJoin ab c
```

So we get this built up `EPreprocessedFrom` which contains the `InnerJoin` thing that gets passed to the callback as `ret`, and the `from_` clause, that is (at this point) the binary tree `FromClause` that *should not* contain any `OnClause`s.

```haskell
-- | (Internal) Finish a @JOIN@.
fromFinish
  :: SqlExpr (PreprocessedFrom a)
  -> SqlQuery a
fromFinish (EPreprocessedFrom ret from_) = Q $ do
  W.tell mempty { sdFromClause = [from_] }
  return ret
```

I think that we have a feasible implementation plan:

- Rendered `SqlExpr (Value Bool)` into a `Text`.
- Parse this into the pair of entities that it should refer to.
  - If the parse fails, we should probably throw an error - an on clause should refer to a pair of entities, though it doesn't matter precisely what the pair is, or what the operator is.
- The `SqlExpr (Value Bool)` that gets passed around will be augmented into `(SqlExpr (Value Bool), Text, Text)`, where the two `Text`s correspond to the table names.
- When checking a `FromJoin`, we'll need to check the `Ident`s associated with the `FromStart Ident EntityDef` and only provide the expression to one where the idents match the two `Text` values provided.

I'll try to get this implemented tomorrow.

**Comment by parsonsmatt** [comment]
Fixed by #156
